File: buff/classes/472/pub/ch1/Penguin.cc
// g++ -g -Wall -o Penguin Penguin.cc && ./Penguin

#include <iostream>

using namespace std;

class Bird {
public:
  virtual void fly()=0;
  virtual void hop()=0;
};

class Penguin : public Bird {
private:
  void fly() { cout << "can't fly" << endl; }
public:
  void hop() { cout << "hop" << endl; }
};

int main() {
  Penguin* p=new Penguin();
  p->hop();
  // p->fly();
  // Bird* b=new Penguin();
  // b->hop();
  // b->fly(); // huh?
}

File: buff/classes/472/pub/ch1/Penguin.java
// javac Penguin.java && java Penguin

interface Bird {
  void fly();
  void hop();
}

public class Penguin implements Bird {

  // private                    // illegal
  public                        // legal
    void fly()       { System.out.println("can't fly"); }
  public void hop()  { System.out.println("hop"); }

  public static void main(String[] args) {
    Penguin h=new Penguin();
    // Bird h=new Penguin();
    h.hop();
    h.fly();
  }

}
